Aplib.Core.Tests/Desire/GoalStructureTests.cs

```diff
@@ -267,6 +267,36 @@ public void RepeatGoalStructure_WhenGoalStructureHasSucceeded_ShouldSucceed()
         currentGoal.Should().Be(goal.Object);
     }
 
+    [Fact]
+    public void RepeatingGoalStructures_WhenSequenced_ShouldNotBeFinished()
+    {
+        // Arrange
+        Mock<IGoal<IBeliefSet>> goal = new();
+        goal.Setup(g => g.GetStatus(It.IsAny<IBeliefSet>())).Returns(CompletionStatus.Success);
+
+        PrimitiveGoalStructure<IBeliefSet> primitiveGoalStructure = new(goal.Object);
+
+        Mock<IGoalStructure<IBeliefSet>> intermediaryGoalStructureMock = new();
+        intermediaryGoalStructureMock.Setup(g => g.Status).Returns(CompletionStatus.Unfinished);
+
+        SequentialGoalStructure<IBeliefSet> sequentialGoalStructure =
+            new(primitiveGoalStructure, intermediaryGoalStructureMock.Object, primitiveGoalStructure);
+
+        // Act
+        sequentialGoalStructure.UpdateStatus(Mock.Of<IBeliefSet>());
+        intermediaryGoalStructureMock.Setup(g => g.Status).Returns(CompletionStatus.Success);
+        goal.Setup(g => g.GetStatus(It.IsAny<IBeliefSet>())).Returns(CompletionStatus.Unfinished);
+        sequentialGoalStructure.UpdateStatus(Mock.Of<IBeliefSet>());
+
+        IGoal<IBeliefSet> currentGoal = sequentialGoalStructure.GetCurrentGoal(Mock.Of<IBeliefSet>());
+
+        // Assert
+        currentGoal.Should().Be(goal.Object);
+        intermediaryGoalStructureMock.Object.Status.Should().Be(CompletionStatus.Success);
+        primitiveGoalStructure.Status.Should().Be(CompletionStatus.Unfinished);
```

Comment:
By using a Mock, you're now not checking our Goal class. You defined the return of Status as CompletionStatus.Unfinished, so it's not testing the actual Goal behavior. The behavior we would like to have tested, is that the actual Goal class resets its status before being evaluated for the second time.

Reply:
The test is not meant to check the Goal behaviour, but the behaviour of the `GoalStructure`. It is impossible to check the status of the goal class, as it will always be behind a `PrimitveGoalStructure`.

Reply:
Ahh okay. In that case I think we need to test it for the Goal class in a separate test, as now there's no guarantee that Goal has the behavior that the Moqs have here. So we should test in another test that Goal has that exact behavior of re-evaluating the status.

We could test this by calling GetStatus two times, where the first time the predicate returns Success, and the second time the predicate returns Failure.

Reply:
Nevermind, there is a test for this already!